src/charm.py

```diff
@@ -1,104 +1,398 @@
+            if pgb_container.pull(USERLIST_PATH).read() == userlist:
+                logger.info("updated userlist does not modify existing pgbouncer userlist")
+                return
+        except FileNotFoundError:
+            # there is no existing userlist.txt file, so carry on and add one.
+            pass
+
+        pgb_container.push(
+            USERLIST_PATH,
+            userlist,
+            user=self._pgbouncer_user,
+            permissions=0o600,
+            make_dirs=True,
+        )
+        logging.info("pushed new userlist to pgbouncer container")
+
+        if reload_pgbouncer:
+            self._reload_pgbouncer()
+
+    def _generate_userlist(self, users: Dict[str, str]) -> str:
+        """Generate userlist.txt from the given dictionary of usernames:passwords.
+
+        Params:
+            users: a dictionary of usernames and passwords
+        Returns:
+            A multiline string, containing each pair of usernames and passwords separated by a
+            space, one pair per line.
+        """
+        return "\n".join([f'"{username}" "{password}"' for username, password in users.items()])
+
+    def _get_users_from_charm_config(self, users=None) -> Dict[str, str]:
+        """Imports users from charm config and generates passwords when necessary.
+
+        TODO it looks as though multiple types of user exist in pgbouncer config, such as
+        stats_users - presumably the details for these users are also stored here. If there are
+        users that aren't listed as either admin or stats-only users, would their information be
+        available here also?
+
+        Params:
+            users: an existing dictionary of usernames and passwords
+        Returns:
+            the `users` dictionary, with usernames from charm config and corresponding generated
+            passwords appended.
+        """
+        if users is None:
+            users = self._get_userlist_from_container()
+
+        for admin_user in self.config["pgb_admin_users"].split(","):
+            # TODO add username validation
+            if admin_user not in users or users[admin_user] is None:
+                users[admin_user] = self._generate_password()
+        return users
+
+    def _get_userlist_from_container(self) -> Dict[str, str]:
+        """Parses container userlist to a dict, for use in charm.
+
+        This parses the userlist.txt stored on the container into a dictionary of strings
+        Returns:
+            users: a dictionary of usernames and passwords
+        """
+        pgb_container = self.unit.get_container(self._pgbouncer_service)
+        parsed_userlist = {}
+        try:
+            userlist = pgb_container.pull(USERLIST_PATH).read()
+        except FileNotFoundError:
+            # there is no existing userlist.txt file, so return nothing
+            return {}
+
+        for line in userlist.split("\n"):
+            username, password = line.replace('"', "").split(" ")
```

Comment:
Seems like that if there is an empty users list file, this line will lead to a `ValueError: not enough values to unpack (expected 2, got 1)` in the `on_install` hook. Probably we can give some more info to the user which is deploying the charm, or handle (here or in the container) the absence of initial users list.